torch/_dynamo/variables/builtin.py

```diff
@@ -1214,10 +1214,41 @@ def _grad_changed(old, new):
                                 grapharg.example.grad = torch.zeros(
                                     grad_shape_specialized, device=new_grad.device
                                 )
-                            else:
-                                grapharg.example.grad = None
-                        return VariableBuilder(tx, source)(grapharg.example.grad)
 
+                        # Check if grad changed via direct setting, such as an inplace op, see
+                        # https://github.com/pytorch/pytorch/pull/112811
+                        if _grad_changed(old_grad, new_grad):
+                            _apply_grad()
+                        else:
+                            # Check if grad changed via attr mutation
+                            if (
+                                obj.mutable_local
+                                and obj.mutable_local
+                                in tx.output.side_effects.store_attr_mutations
+                            ):
+                                mutations = tx.output.side_effects.store_attr_mutations[
+                                    obj.mutable_local
+                                ]
+                                if "grad" in mutations:
+                                    if mutations["grad"] is not None:
+                                        new_grad = (
+                                            mutations["grad"]
+                                            .as_proxy()
+                                            .node.meta["example_value"]
+                                        )
+                                    else:
+                                        new_grad = None
+                                if _grad_changed(old_grad, new_grad):
+                                    # _apply_grad()
+                                    # TODO(voz): We want to support this properly, but DO NOT want
+                                    # setattr in the graph - we need to sit down here and think about
+                                    # a design. Potentially a set_grad op?
+                                    unimplemented(
```

Comment:
Wouldn't we need to do this graph break in the `setattr` op?  This looks like the `getattr`?  Which would result in the `setattr` still being in the graph.

Reply:
Well, the setattr wont be in the graph, because we treat .grad setting via side effects - it's tracked separately. The problem is from when we access it in the graph after setting it. But you are right that we should probably graph break on the setattr and call it a day.

Reply:
But if we graph break here, the grad setting will still be in the side effects and we will codegen it.